### PR TITLE
Auth 2fa

### PR DESCRIPTION
**Pull request by lefoun**
**Changes**
- 2 Factor Authentication
- Local Authentication made a comeback (now passes all tests and password policy is updated).
- 2e2 and integration tests completed for both of the above.
- Updated swagger to account for the changes made in the branch.
- Refactor code to make it more readable and more consistent.

### CODE REVIEW

File: srcs/client/src/App.jsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import "./App.css";
+import { BrowserRouter, Routes, Route, Navigate } from "react-router-dom";
+import { useEffect, useState } from "react";
+
+const App = () => {
+  const [user, setUser] = useState(null);
+  const [loginOrRegister, setAuthState] = useState(null);
+  useEffect(() => {
+    const getUser = () => {
+      fetch("http://127.0.0.1:4200/auth/success", {
```
Comment (twagger) [inline srcs/client/src/App.jsx:16]: About this address I am not sure about the effect in production as we do not expose api server to the client. We may have to think about it soon.

File: srcs/server/prisma/migrations/20220827134504_/migration.sql [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+*/
+-- AlterTable
+ALTER TABLE "Credentials" DROP COLUMN "two_fa_activated",
+DROP COLUMN "two_fa_secret",
+ADD COLUMN     "twoFactorActivated" BOOLEAN NOT NULL DEFAULT false,
+ADD COLUMN     "twoFactorSecret" TEXT;
+
+-- AlterTable
+ALTER TABLE "User" DROP COLUMN "two_fa_activated",
+DROP COLUMN "two_fa_secret";
```
Comment (twagger) [inline srcs/server/prisma/migrations/20220827134504_/migration.sql:18]: You can name the migrations using the command :
`npx prisma migrate dev --name 2fa_on_credentials`

for example. It is usefull as we will have a lot of migrations at the end of the project to quicky identify them.

File: srcs/server/src/auth/exceptions/index.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export * from "./badCredentials.exception";
+export * from "./shouldBeLogged-in.exception";
+export * from "./internalAuth.exception";
+export * from "./userAlreadyRegistered.exception";
+export * from "./credentialsTaken.exception";
```
Comment (twagger) [inline srcs/server/src/auth/exceptions/index.ts:5]: I like this idea of having an index for exception to ease the import. I will apply it on my code.

**Review by twagger — APPROVED** [review]
Ok for me

File: srcs/client/src/App.jsx (hunk 1 above)
Comment (lefoun) [reply srcs/client/src/App.jsx:16]: I'm not sure I follow. How would the client make requests to start a new game for example or to get user data ?

But yeah this isn't the cleanest way of doing it and I think the team is going to do something else. This was just to test the flow.

File: srcs/server/prisma/migrations/20220827134504_/migration.sql (hunk 2 above)
Comment (lefoun) [reply srcs/server/prisma/migrations/20220827134504_/migration.sql:18]: I think it's a better idea to name them as well!

But the migration was done automatically when I launched the makefile.

Next time I'll make sure to migrate --name before relaunching a docker compose.

File: srcs/client/src/App.jsx (hunk 1 above)
Comment (twagger) [reply srcs/client/src/App.jsx:16]: In production the website will be served by a web server (nginx for example). As a user, I should not have to access a backend directly to process my queries. I think the proper way to do it in production will be with nginx proxying backend traffic with a specific location (/api for example).

**Review by Mastodoonte — APPROVED** [review]
All good, need to merge front work on master soon

**Review by fmonbeig — APPROVED** [review]

**Pull request merged**